opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,676 @@
+}
+
+template<typename ActionT>
+void FollowingServer::getPreemptedGoalIfRequested(
+  typename std::shared_ptr<const typename ActionT::Goal> goal,
+  const typename nav2::SimpleActionServer<ActionT>::SharedPtr & action_server)
+{
+  if (action_server->is_preempt_requested()) {
+    goal = action_server->accept_pending_goal();
+  }
+}
+
+template<typename ActionT>
+bool FollowingServer::checkAndWarnIfCancelled(
+  typename nav2::SimpleActionServer<ActionT>::SharedPtr & action_server,
+  const std::string & name)
+{
+  if (action_server->is_cancel_requested()) {
+    RCLCPP_WARN(get_logger(), "Goal was cancelled. Cancelling %s action", name.c_str());
+    return true;
+  }
+  return false;
+}
+
+template<typename ActionT>
+bool FollowingServer::checkAndWarnIfPreempted(
+  typename nav2::SimpleActionServer<ActionT>::SharedPtr & action_server,
+  const std::string & name)
+{
+  if (action_server->is_preempt_requested()) {
+    RCLCPP_WARN(get_logger(), "Goal was preempted. Cancelling %s action", name.c_str());
+    return true;
+  }
+  return false;
+}
+
+void FollowingServer::followObject()
+{
+  std::lock_guard<std::mutex> lock(dynamic_params_lock_);
+  action_start_time_ = this->now();
+  rclcpp::Rate loop_rate(controller_frequency_);
+
+  auto goal = following_action_server_->get_current_goal();
+  auto result = std::make_shared<FollowObject::Result>();
+
+  if (!following_action_server_ || !following_action_server_->is_server_active()) {
+    RCLCPP_DEBUG(get_logger(), "Action server unavailable or inactive. Stopping.");
+    return;
+  }
+
+  if (checkAndWarnIfCancelled<FollowObject>(following_action_server_, "follow_object")) {
+    following_action_server_->terminate_all();
+    return;
+  }
+
+  getPreemptedGoalIfRequested<FollowObject>(goal, following_action_server_);
+  num_retries_ = 0;
+  static_timer_initialized_ = false;
+
+  try {
+    auto pose_topic = goal->pose_topic;
+    auto target_frame = goal->tracked_frame;
+    if (target_frame.empty()) {
+      if (pose_topic.empty()) {
+        RCLCPP_ERROR(get_logger(),
+          "Both pose topic and target frame are empty. Cannot follow object.");
+        result->error_code = FollowObject::Result::FAILED_TO_DETECT_OBJECT;
+        result->error_msg = "No pose topic or target frame provided.";
+        following_action_server_->terminate_all(result);
+        return;
+      } else {
+        RCLCPP_INFO(get_logger(), "Subscribing to pose topic: %s", pose_topic.c_str());
+        dynamic_pose_sub_ = create_subscription<geometry_msgs::msg::PoseStamped>(
+          pose_topic,
+          [this](const geometry_msgs::msg::PoseStamped::SharedPtr pose) {
+            detected_dynamic_pose_ = *pose;
+            RCLCPP_DEBUG(get_logger(), "Received dynamic pose");
+          },
+          nav2::qos::StandardTopicQoS(1));  // Only want the most recent pose
+      }
+    } else {
+      RCLCPP_INFO(get_logger(), "Following frame: %s instead of pose", target_frame.c_str());
+    }
```

Comment:
I'm a little unclear on this block. I don't see where we set to use either the topic's pose information or extract the pose information from the tracked frame using TF. It seems like the `detected_dynamic_pose_` is always used in `getRefinedPose`, but only 1 of 2 (?) places use `getRefinedPose` or `getFramePose`.